### PR TITLE
Uncouple CacheDirectory from SearchableSnapshotDirectory

### PR DESCRIPTION
**Pull request by tlrx**
Today `CacheDirectory` is implemented as a `FilterDirectory` that caches files locally while delegating the read operations to `SearchableSnapshotDirectory`.

This was very useful to separate concerns like caching Lucene files on disk from reading Lucene files from a blob store repository, but it comes with additional complexity:
- `IndexInput` are buffered within each directory, making it difficult to understand the reading pattern
- in case of cache evictions, the `CacheDirectory` attempts to directly read _N_ bytes from the `IndexInput` but because of the `SearchableSnapshotDirectory` read ahead it might in fact download more bytes than needed
- the direct read bytes stat does not exactly reflect the bytes read by the underlying `BlobContainer`
- changes like #53492 are a bit more complex than necessary

This pull request is a first step forward merging `CacheDirectory` into `SearchableSnapshotDirectory`. It changes the cache directory so that it does not rely on the searchable snapshot directory anymore and instead read the bytes directly from the `BlobContainer`. It also adds two more base classes that group common class attributes for directories and index inputs.

### CODE REVIEW

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/BaseSearchableSnapshotDirectory.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshot.FileInfo;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public abstract class BaseSearchableSnapshotDirectory extends BaseDirectory {
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/BaseSearchableSnapshotDirectory.java]: This abstract class has been introduced in order to contain common attributes of existing directories and in order to unify the existing constructors. It should become a concrete `SearchableSnapshotDirectory` when the cache logic and the searchable snapshot logic will be merged together.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java [hunk 2]
```diff
@@ -158,11 +79,13 @@ public static Directory create(RepositoriesService repositories,
             SNAPSHOT_SNAPSHOT_ID_SETTING.get(indexSettings.getSettings()));
         final BlobStoreIndexShardSnapshot snapshot = blobStoreRepository.loadShardSnapshot(blobContainer, snapshotId);
 
-        Directory directory = new SearchableSnapshotDirectory(snapshot, blobContainer);
+        final Directory directory;
         if (SNAPSHOT_CACHE_ENABLED_SETTING.get(indexSettings.getSettings())) {
             final Path cacheDir = shardPath.getDataPath().resolve("snapshots").resolve(snapshotId.getUUID());
-            directory = new CacheDirectory(directory, cache, cacheDir, snapshotId, indexId, shardPath.getShardId(),
+            directory = new CacheDirectory(snapshot, blobContainer, cache, cacheDir, snapshotId, indexId, shardPath.getShardId(),
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java]: With this pull request, CacheDirectory reads blobs using the BlobContainer and does not need to wrap the SearchableSnapshotDirectory anymore.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java [hunk 3]
```diff
@@ -61,17 +60,21 @@
     private static final long NO_SEQUENTIAL_READ_OPTIMIZATION = 0L;
 
 
-    SearchableSnapshotIndexInput(final BlobContainer blobContainer, final FileInfo fileInfo, long sequentialReadSize, int bufferSize) {
-        this("SearchableSnapshotIndexInput(" + fileInfo.physicalName() + ")", blobContainer, fileInfo, 0L, 0L, fileInfo.length(),
+    SearchableSnapshotIndexInput(
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java]: I'll revert this spotless formatting

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheDirectory.java [hunk 4]
```diff
@@ -39,7 +43,7 @@
 /**
  * {@link CacheDirectory} uses a {@link CacheService} to cache Lucene files provided by another {@link Directory}.
  */
-public class CacheDirectory extends FilterDirectory {
+public class CacheDirectory extends BaseSearchableSnapshotDirectory {
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheDirectory.java]: This is the main purpose of this pull request :)

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheDirectory.java [hunk 5]
```diff
@@ -287,23 +293,21 @@ int readCacheFile(FileChannel fc, long end, long position, byte[] buffer, int of
...
             assert assertFileChannelOpen(fc);
-            final byte[] copyBuffer = new byte[Math.toIntExact(Math.min(COPY_BUFFER_SIZE, end - start))];
+            final long length = end - start;
+            final byte[] copyBuffer = new byte[Math.toIntExact(Math.min(COPY_BUFFER_SIZE, length))];
             logger.trace(() -> new ParameterizedMessage("writing range [{}-{}] to cache file [{}]", start, end, cacheFileReference));
 
             int bytesCopied = 0;
-            try (IndexInput input = in.openInput(cacheFileReference.getFileName(), ioContext)) {
+            final long startTimeNanos = currentTimeNanosSupplier.getAsLong();
+            try (InputStream input = openInputStream(start, length)) {
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheDirectory.java]: There is an important change here: by reading the range to cache on disk using the `BlobContainer` only the required bytes are downloaded in a (usually) single request (unless the range spans over multiple blob parts) and there is no read ahead engaged here.

**Comment by tlrx** [comment]
@elasticmachine update branch

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java [hunk 6]
```diff
@@ -50,95 +42,24 @@
  * shard files and what it stored in the snapshot the {@link BlobStoreIndexShardSnapshot} is used to map a physical file name as expected by
  * Lucene with the one (or the ones) corresponding blob(s) in the snapshot.
  */
-public class SearchableSnapshotDirectory extends BaseDirectory {
-
-    private final BlobStoreIndexShardSnapshot snapshot;
-    private final BlobContainer blobContainer;
+public class SearchableSnapshotDirectory extends BaseSearchableSnapshotDirectory {
```
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java]: This is only used if `index.store.snapshot.cache.enabled` is false. Maybe we can remove that setting and this whole class?

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java (hunk 3 above)
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java]: I for one welcome our new spotless overlords :ant:

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheDirectory.java (hunk 5 above)
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheDirectory.java]: :+1:

**Review by DaveCTurner — COMMENTED** [review]
Looks good @tlrx, but I think we can go further. I think we don't need to run things with the cache completely disabled any more, so we can drop the `index.store.snapshot.cache.enabled` setting, which means `SearchableSnapshotDirectory` is no longer needed, which means each base class has just one concrete implementation and so they can be merged together.

I'm ok with reviewing all that in one PR, I think, especially if the steps are mostly in separate commits. WDYT?

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/BaseSearchableSnapshotIndexInput.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        FileInfo fileInfo,
+        IOContext context,
+        int bufferSize
+    ) {
+        this(resourceDesc, blobContainer, fileInfo, context);
+        setBufferSize(bufferSize);
+    }
+
+    protected InputStream openInputStream(final long position, final long length) throws IOException {
+        // TODO move this at the Directory level
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/BaseSearchableSnapshotIndexInput.java]: We discussed this in another channel and decided to leave it for a followup in which we consolidate how we choose between all the different `IndexInput` implementations.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java (hunk 6 above)
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java]: We discussed this in another channel and agreed to defer this to a followup.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java (hunk 2 above)
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java]: :+1:

**Review by DaveCTurner — DISMISSED** [review]
LGTM

**Review by DaveCTurner — COMMENTED** [review]
sorry, LGTM'd it too soon, there's a couple of places where I think we can assert something's not called.

File: x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheBufferedIndexInputTests.java [hunk 8]
```diff
@@ -56,145 +69,110 @@ public void testRandomReads() throws IOException {
...
+        CountingBlobContainer(BlobContainer in, int rangeSize) {
             super(in);
             this.rangeSize = rangeSize;
         }
 
         @Override
-        public IndexInput openInput(String name, IOContext context) throws IOException {
-            return new CountingIndexInput(this, super.openInput(name, context), rangeSize);
+        public InputStream readBlob(String name) throws IOException {
+            return new CountingInputStream(this, super.readBlob(name), getBlobLength(name), rangeSize);
```
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheBufferedIndexInputTests.java]: Sure

**Comment by tlrx** [comment]
Thanks @DaveCTurner - I've applied your feedback and also pushed a small change in `BlobContainerWrapper` that didn't make it in the previous commit.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by tlrx** [comment]
Thanks David!

**Pull request merged**